pkg/services/ngalert/api/api_ruler_validation.go

```diff
@@ -59,46 +61,6 @@ func validateRuleNode(
 		return nil, fmt.Errorf("alert rule title is too long. Max length is %d", store.AlertRuleMaxTitleLength)
 	}
 
-	noDataState := ngmodels.NoData
-	if ruleNode.GrafanaManagedAlert.NoDataState == "" && canPatch {
-		noDataState = ""
-	}
-
-	if ruleNode.GrafanaManagedAlert.NoDataState != "" {
-		noDataState, err = ngmodels.NoDataStateFromString(string(ruleNode.GrafanaManagedAlert.NoDataState))
-		if err != nil {
-			return nil, err
-		}
-	}
-
-	errorState := ngmodels.AlertingErrState
-
-	if ruleNode.GrafanaManagedAlert.ExecErrState == "" && canPatch {
-		errorState = ""
-	}
-
-	if ruleNode.GrafanaManagedAlert.ExecErrState != "" {
-		errorState, err = ngmodels.ErrStateFromString(string(ruleNode.GrafanaManagedAlert.ExecErrState))
-		if err != nil {
-			return nil, err
-		}
-	}
-
-	if len(ruleNode.GrafanaManagedAlert.Data) == 0 {
-		if canPatch {
-			if ruleNode.GrafanaManagedAlert.Condition != "" {
-				return nil, fmt.Errorf("%w: query is not specified by condition is. You must specify both query and condition to update existing alert rule", ngmodels.ErrAlertRuleFailedValidation)
-			}
-		} else {
-			return nil, fmt.Errorf("%w: no queries or expressions are found", ngmodels.ErrAlertRuleFailedValidation)
-		}
-	} else {
-		err = validateCondition(ruleNode.GrafanaManagedAlert.Condition, ruleNode.GrafanaManagedAlert.Data)
-		if err != nil {
-			return nil, fmt.Errorf("%w: %s", ngmodels.ErrAlertRuleFailedValidation, err.Error())
-		}
-	}
```

Comment:
All this extra logic surrounding `validateCondition` is now shoved inside `validateCondition`. validateCondition now handles all these cases and can be directly called from any path.

Reply:
You're referring to `validateAlertingRule` here right?

Reply:
I mean the logic surrounding `validateCondition` (if canPatch { ...). It's put inside validateCondition now, so code can just call `validateCondition` straight up without needing to worry about this extra stuff that's also validating part of the condition.

Then _on top of that_, the call to validateCondition got put inside validateAlertingRule. but, the comment was referring to the surrounding logic being put inside validateCondition. That's why you don't see this whole block copied inside `validateAlertingRule`, just the call to validateCondition and nothing else.